Create a shared lib with C API

MikeHolman opened the pull request:
This change adds a new shared library om with a C API to use as an alternative to the executable.

This is motivated by the desire to avoid having to serialize onnx graphs to disk before calling onnx-mlir.

The library currently only has two methods: CompileFromFile and CompileFromArray, where CompileFromFile acts in roughly the same way as onnx-mlir.exe does today, compiling from a file containing MLIR or an ONNX protobuf. CompileFromArray compiles from an in-memory ONNX protobuf array.

AlexandreEichenberger commented:
I asked @gongsu832 to review it as he has experience with our interfaces, and @chentong319 because he does a lot of our front-end work.

I have a high level question: are you planning to have a test of this new interface? If we are adding new functionality, it would be great to ensure that it works through our CIs.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/onnx/onnx-mlir?pullRequest=894) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/onnx/onnx-mlir?pullRequest=894) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: AlexandreEichenberger<br/>:x: MikeHolman<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/onnx/onnx-mlir?pullRequest=894) it.</sub>

gongsu832 commented:
A few comments.

- I'm not sure we need another lib for this. Can't the functions be part of `MainUtils`?
- None of the other sources uses dynamically generated export headers so using it for just this lib doesn't really buy much and makes the mixed-style source harder to maintain. I'm also in general wary of dynamically generated code as they typically contain a lot of dead code that never get exercised and are typically not for human to read.
- The naming for the targets (e.g., `om`), libraries (e.g., `OMLibrary`), etc. is rather arbitrary. They need to be more meaningful and specific. They should also follow existing naming style. For example, for files under the src directory, we don't prefix them with `OM`.

MikeHolman commented:
The reason to have a shared library is because we have a consumer team who uses onnx-mlir.exe and adding onnx-mlir as a build dependency is prohibitively expensive. With a shared library we can simply give them the dll and not require them to add onnx-mlir and all its dependencies (in particular LLVM/MLIR) to their build.

I've removed the dynamic export header.

I didn't think the names were arbitrary, this is meant as a general purpose alternative to onnx-mlir, so I kept the name general. But in any case I've renamed and moved the files to be more clear that this is specifically a compiler API. Let me know what you think of the changes.

gongsu832 commented:
@jenkins-droid test this please

caoimhinuibrian commented:
@jenkins-droid test this please

gongsu832 commented:
@MikeHolman Yes I think this looks much better. We can keep them separate for now and make further improvements and decide whether they should be merged or not according to how things evolve. Thanks.

gongsu832 reviewed (approved):
LGTM

AlexandreEichenberger commented on src/Compiler/OnnxMlirCompiler.cpp:
```diff
@@ -0,0 +1,58 @@
+    const char *mcpu, const char *mtriple) {
+  mlir::OwningModuleRef module;
+  mlir::MLIRContext context;
+  std::string mcpuString;
```
Suggestion: could we have the code that is in common between the two CompileFromFile and CompileFromArray in a common function, say SetCompileContext, handling them cpu, mtriple, dialect registration?

AlexandreEichenberger commented on include/OnnxMlirCompiler.h:
```diff
@@ -0,0 +1,69 @@
+ *  @param mtriple Optional target architecture
+ *  @return 0 on success or non-zero error code on failure
+ */
+ONNX_MLIR_EXPORT int CompileFromFile(const char *inputFilename,
```
All of the names in the ONNX MLIR runtime are prefaced with "om". I suggest we do the same here.

AlexandreEichenberger commented on src/Compiler/OnnxMlirCompiler.cpp:
```diff
@@ -0,0 +1,45 @@
+#include "OnnxMlirCompiler.h"
+#include "CompilerUtils.hpp"
+
+void SetCompileContext(
```
You can check elsewhere, but we are also trying to use small cap for the function, and we like them to be verbs, so `setCompileContext` would be in the spirit of things

AlexandreEichenberger commented:
@jenkins-droid test this please
